### ISSUE TITLE
Upgrade/downgrade domain behaviour on poweroff

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#675]
Originally reported of Google Code with ID 619.

```
In Ganeti 2.10, the Xen hypervisor is responsible for cleaning up the domain once that domain is shutdown, whether the shutdown is issued by Ganeti or by a user.

In Ganeti 2.11 (currently master branch), the Xen hypervisor is no longer responsible for this cleanup, and instead Ganeti is. As a result, Ganeti needs to clean up the domain after a shutdown from Ganeti itself, or clean up the domain after a user shutdown through the watcher.

This issue focuses on the upgrade/downgrade procedure and the possible troubling pairs of Ganeti version and Xen domain configuration.

Consider an instance created with 2.10 followed by an upgrade of Ganeti from 2.10 to 2.11. In order to stop the instance, the user will issue 'gnt-instance stop'. What will happen is that the instance will suffer an additional and unnecessary 'xm destroy' command from the Ganeti 2.11 codebase, which will produce an error because this command is trying to remove a domain that was already cleanup by Xen, given that is the behaviour that is part of the 2.10 instance configuration. This is easy to fix simply by changing 2.11 to make sure the instance is still running before issuing the final 'xm destroy' command.

Consider now the reverse situation: an instance created with 2.11 followed by a downgrade of Ganeti from 2.11 to 2.10. The 2.11 instance was configured such that the expected cleanup behaviour comes from Ganeti and not from Xen. However, Ganeti 2.10 has a shutdown retry mechanism in place that will try to shutdown a domain several times and, if unsuccessful, it will destroy the domain. This mechanism should be enough to handle shutdown of domains configured for 2.11. 
```
Originally added on 2013-11-22 12:05:47 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-11-22 12:06:18 +0000 UTC.

Added Labels: Type-Defect

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-11-22 13:55:30 +0000 UTC.

Added to Milestone: Release2.11

**Comment by Ganeti-Issues-Migrator** [comment]
```
Commit in stable-2.8.

From 364c350ff121920c7d82716c85178745ea4e6822 Mon Sep 17 00:00:00 2001
From: "Jose A. Lopes" <jabolopes@google.com>
Date: Mon, 2 Dec 2013 12:41:33 +0100
Subject: [PATCH] Xen handle domain shutdown

Update Xen backend to properly recognize when a domain has been
shutdown by the user and to properly cleanup a shutdown domain when
Ganeti requests Xen to stop this domain.

Partial cherry-pick from 9d22cc90609e3ee8f0f2b34b793a3daced3c0e61

Signed-off-by: Jose A. Lopes <jabolopes@google.com>
Reviewed-by: Michele Tartara <mtartara@google.com>
```
Originally added on 2014-01-23 08:58:33 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Merge into stable-2.9.

From: Thomas Thrainer <thomasth@google.com>
Date: Mon, 16 Dec 2013 09:49:55 +0000 (+0100)
Subject: Merge branch 'stable-2.9' into stable-2.10
X-Git-Tag: v2.10.0rc1~4
X-Git-Url: http://git.ganeti.org/?p=ganeti.git;a=commitdiff_plain;h=c486fb6cc69da18fea383f2194dacb5ef6f8b94a

Merge branch 'stable-2.9' into stable-2.10

* stable-2.9
  Bump revision for 2.9.2
  Update NEWS for 2.9.2 release
  Pass hvparams to GetInstanceInfo
  Adapt parameters that moved to instance variables
  Avoid lines longer than 80 chars
  SingleNotifyPipeCondition: don't share pollers
  KVM: use custom KVM path if set for version checking
* stable-2.8
  Version bump for 2.8.3
  Update NEWS for 2.8.3 release
  Support reseting arbitrary params of ext disks
  Allow modification of arbitrary params for ext
  Do not clear disk.params in UpgradeConfig()
  SetDiskID() before accepting an instance
  Lock group(s) when creating instances
  Fix job error message after unclean master shutdown
  Add default file_driver if missing
  Update tests
  Xen handle domain shutdown
  Fix evacuation out of drained node
  Refactor reading live data in htools
  master-up-setup: Ping multiple times with a shorter interval
  Add a packet number limit to "fping" in master-ip-setup
  Fix a bug in InstanceSetParams concerning names
  build_chroot: hard-code the version of blaze-builder
  Fix error printing
  Allow link local IPv6 gateways
  Fix NODE/NODE_RES locking in LUInstanceCreate
  eta-reduce isIpV6
  Ganeti.Rpc: use brackets for ipv6 addresses
  Update NEWS file with socket permission fix info
  Fix socket permissions after master-failover

Conflicts:
	NEWS
	configure.ac
	devel/build_chroot
	lib/constants.py
	src/Ganeti/Rpc.hs

Resolution:
    NEWS: take both additions
    configure.ac: ignore version bump
    constants.py: move constants to Constants.hs
    instance_migration.py: Remove call to SetDiskID(...), it has been removed in 2.10
    instance_unittest.py: Adapt test to new logic in LU
    Rest: trivial

Signed-off-by: Thomas Thrainer <thomasth@google.com>
Reviewed-by: Michele Tartara <mtartara@google.com>
```
Originally added on 2014-01-23 08:59:27 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Merge into stable-2.11.

From: Klaus Aehlig <aehlig@google.com>
Date: Thu, 19 Dec 2013 10:06:22 +0000 (+0100)
Subject: Merge branch 'stable-2.10' into master
X-Git-Url: http://git.ganeti.org/?p=ganeti.git;a=commitdiff_plain;h=f3ac6f36385141cdd1a673872bda85f24d1be0fc

Merge branch 'stable-2.10' into master

* stable-2.10
  Version bump for 2.10.0~rc1
  Update NEWS for 2.10.0 rc1 release
  Fix pylint 0.26.0/Python 2.7 warning
  Update INSTALL and devnotes for 2.10 release
* stable-2.9
  Bump revision for 2.9.2
  Update NEWS for 2.9.2 release
  Pass hvparams to GetInstanceInfo
  Adapt parameters that moved to instance variables
  Avoid lines longer than 80 chars
  SingleNotifyPipeCondition: don't share pollers
  KVM: use custom KVM path if set for version checking
* stable-2.8
  Version bump for 2.8.3
  Update NEWS for 2.8.3 release
  Support reseting arbitrary params of ext disks
  Allow modification of arbitrary params for ext
  Do not clear disk.params in UpgradeConfig()
  SetDiskID() before accepting an instance
  Lock group(s) when creating instances
  Fix job error message after unclean master shutdown
  Add default file_driver if missing
  Update tests
  Xen handle domain shutdown
  Fix evacuation out of drained node
  Refactor reading live data in htools
  master-up-setup: Ping multiple times with a shorter interval
  Add a packet number limit to "fping" in master-ip-setup
  Fix a bug in InstanceSetParams concerning names
  build_chroot: hard-code the version of blaze-builder
  Fix error printing
  Allow link local IPv6 gateways
  Fix NODE/NODE_RES locking in LUInstanceCreate
  eta-reduce isIpV6
  Ganeti.Rpc: use brackets for ipv6 addresses
  Update NEWS file with socket permission fix info
  Fix socket permissions after master-failover

Conflicts:
	NEWS
	devel/build_chroot
	lib/cmdlib/instance.py
	lib/hypervisor/hv_xen.py
	lib/jqueue.py
	src/Ganeti/Luxi.hs
	tools/cfgupgrade
Resolution:
	- tools/cfgupgrade: ignore downgrade changes from 2.10
	- NEWS: take both changes
	- devel/build_chroot: both changes differed only in indentation;
	    use indentation from master
	- lib/hypervisor/hv_xen.py: manually apply fd201010 and 70d8491f to
	    the conflicting hunks from stable-2.10
	- lib/jqueue.py: manually apply 9cbcb1be to the conflicting hunk from
	    master
Semanitcal conflicts:
	- configure.ac: undo revision bump to ~rc1
	- lib/query.py: manually merge the two independently added
	    functions _GetInstAllNicVlans

Signed-off-by: Klaus Aehlig <aehlig@google.com>
Reviewed-by: Helga Velroyen <helgav@google.com>
```
Originally added on 2014-01-23 09:00:25 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-04-28 09:06:47 +0000 UTC.

Changed State: Released
